### ISSUE TITLE
Incorrect cookie parsing

### ISSUE THREAD

**Issue by ZetaTwo** [embedthis/goahead#307]
### Environment

1. O/S Version: Ubuntu 21.04

2. GoAhead Version: 5.1.5 (built from master)

3. Hardware: HP Zbook laptop (irreleant)

### Build

1. If built from source, what was your configure command line:

```
./configure
```

### Problem Description:

websParseCookie() uses a very naive parser to fetch cookie values from a header. This enables a maliciously crafted cookie to affect the extracted value of another cookie. I have attached a PoC with a very simple handler that simply reflects the values of two cookies: "cookie_a" and "cookie_b".

Consider the two cURL requests below, in the first one, cookie_a is before cookie_b in the cookie header and no issue occurs. In the second request cookie_b is before cookie_a in the request and the parser incorrectly extracts the value thats inside cookie_b and treats it as the value for cookie_a.

In the attached PoC I have also included a simple html page which will perform these requests from JS to further show the issue but note that the ordering of the cookies is browser specific so it may or may not work depending on your environment. At worst this bug can lead to security issues where an attacker manages to control the value of one cookie and uses it to alter the servers interpretation of any other cookie sent.

````
$ curl -v http://localhost:8000/poc -H 'Cookie: cookie_a=safe_value; cookie_b= cookie_a=evil_value;'
*   Trying 127.0.0.1:8000...
* Connected to localhost (127.0.0.1) port 8000 (#0)
> GET /poc HTTP/1.1
> Host: localhost:8000
> User-Agent: curl/7.74.0
> Accept: */*
> Cookie: cookie_a=safe_value; cookie_b= cookie_a=evil_value;
> 
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< Date: Sun Oct 24 18:06:09 2021
< Content-Length: 58
< Connection: keep-alive
< X-Frame-Options: SAMEORIGIN
< 
Cookie A: "safe_value"
Cookie B: "cookie_a=evil_value"
````

```
$ curl -v http://localhost:8000/poc -H 'Cookie: cookie_b= cookie_a=evil_value; cookie_a=safe_value'
*   Trying 127.0.0.1:8000...
* Connected to localhost (127.0.0.1) port 8000 (#0)
> GET /poc HTTP/1.1
> Host: localhost:8000
> User-Agent: curl/7.74.0
> Accept: */*
> Cookie: cookie_b= cookie_a=evil_value; cookie_a=safe_value
> 
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< Date: Sun Oct 24 18:06:27 2021
< Content-Length: 57
< Connection: keep-alive
< X-Frame-Options: SAMEORIGIN
< 
Cookie A: "evil_value"
Cookie B: "cookie_a=evil_value"
```

[poc.zip](https://github.com/embedthis/goahead/files/7406005/poc.zip)

**Comment by mobsense** [comment]
Sorry for the delay in getting to this.

Thank you for the detailed report. It was really easy to reproduce and your description was really clear. Thank you.

Couple of comments:

* websParseCookie is an internal API and isn't public. This scopes the issue, but does not invalidate it.
* The parsing logic ignores invalid cookies of the form "cooke= " and steps over them. Then it parses cookies and looks for a semicolon or comma terminator.
* The client totally controls the cookie string, so using trickery to set cookies to arbitrary values can be simply achieved by setting the cookies to the desired "evil" string.

So Net/Net: I don't think this is a security issue. 

Please feel free to push back on this and outline how this could be a security issue.

**Comment by ZetaTwo** [comment]
```
The client totally controls the cookie string, so using trickery to set cookies to arbitrary values can be simply achieved by setting the cookies to the desired "evil" string.
```

The issue arises when a third-party attacker is able to influence the value of one cookie but not the other. Imagine the following scenario:

1. The application uses two cookies `column_order` and `csrf_token`.
2. The `column_order` cookie is a simple cookie that determines which order you want the columns of a table to be rendered. This cookie is updated by making a request to `vuln.com/set_column_order?order=col1,col2,col3` which will then reply back with a `Set-Cookie: column_order=col1,col2,col3` header.
3. The attacker hosts an evil page which the victim visits. The page initiates a request to `vuln.com/set_column_order?order=%20csrf_token=A`
4. The attacker page then initiates a request to `vuln.com/admin/sensitive_action`. This is protected by a CSRF token but due to this cookie parser  error, the attacker knows that the CSRF token will be interpreted as `A` and can set this value in the headers and defeat the CSRF protection.

In short, this can absolutely have security impact in certain situations.

For even more context, this bug and another bug in goahead was used as part of a [security competition](https://ctftime.org/event/1415) last year where you had to hack a website using goahead.

**Comment by mobsense** [comment]
Thanks for the extra detail.

How can an attacker have access to one cookie and not the other?  Surely, the attacker or MITM can modify the entire cookie string?

**Comment by ZetaTwo** [comment]
No this is not a MitM scenario. Initially the attacker does not know anything about the current value of the cookie but if the web application allows a user to set a cookie based on input to the application then the attacker can abuse this to make the user set that cookie to this malicious value.

In my example above the attacker hosts a malicious website evil.com which contains some javascript. The victim visits this website. The javascript, which runs in the context of the victim makes a request to the vulnerable web app to set the cookie value to a value that is known to the attacker. In my example above I called this cookie `column_order`. Note here that this by itself would not be an issue.

The issue arises when you then initiate another request from the malicious page to a sensitive action, such as an admin action. The admin action is protected with CSRF where a post value needs to match a cookie value. Since the attacker can neither see nor modify the value of the `csrf_token` cookie this should be fine.

This is where the issue occurs, because the goahead server will incorrectly parse the `column_order` cookie which the attacker have managed to influence and think that the value of `csrf_token` is `A` which matches what the attacker sends in the form value as well and thus the CSRF protection has been bypassed.

**Comment by mobsense** [comment]
Thank for the details.

Can you check out this alternate parsing of the cookie header that is less tolerant of illegal headers:

It follows the spec more closely and is less tolerant of legacy non-compliant cookie header formulations.

```
PUBLIC char *websParseCookie(Webs *wp, char *name)
{
    char    *buf, *cookie, *end, *key, *tok, *value;

    assert(wp);

    if (wp->cookie == 0 || name == 0 || *name == '\0') {
        return 0;
    }
    buf = sclone(wp->cookie);
    end = &buf[slen(buf)];
    value = 0;

    for (tok = buf; tok < end; ) {
         cookie = stok(tok, ";", &tok);
         key = stok(cookie, "=", &value);
         if (smatch(key, name)) {
             // Remove leading spaces first, then double quotes. Spaces inside double quotes preserved.
             value = sclone(strim(strim(value, " ", WEBS_TRIM_BOTH), "\"", WEBS_TRIM_BOTH));
             break;
         }
    }
    wfree(buf);
    return value;
}

```
